### PR TITLE
Refactor `update_metric`

### PR DESCRIPTION
**Pull request by maxisawesome**
# What does this PR do?

Refactors `huggingface.py`'s `update_metric` to not require checking the metric's type.

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by bmosaicml — APPROVED** [review]
I approve of this design and think it would be the best way to remove the InContextLearning classes from composer. Thanks Max!

**Review by eracah — APPROVED** [review]
LGTM. So some metrics need the batch because some metric configs are stored in the batch? hmmm seems like there should be a cleaner way...

**Review by dakinggg — CHANGES_REQUESTED** [review]
I suggested an alternate design on the doc, blocking until we've discussed a bit.

File: composer/models/huggingface.py [hunk 1]
```diff
@@ -532,12 +532,8 @@ def get_metrics(self, is_train: bool = False) -> Dict[str, Metric]:
         return metrics if metrics else {}
 
     def update_metric(self, batch: Any, outputs: Any, metric: Metric) -> None:
-        if isinstance(metric, InContextLearningQAAccuracy):
-            assert self.labels is not None
+        if getattr(metric, 'requires_batch', False):
```
Comment (dakinggg) [inline composer/models/huggingface.py:535]: attr is named wrong, its `need_batch`

**Review by dakinggg — APPROVED** [review]

**Comment by bmosaicml** [comment]
> LGTM. So some metrics need the batch because some metric configs are stored in the batch? hmmm seems like there should be a cleaner way...

It's not just global configs stored (though that is the case for some metrics, and def not good design), but also batch level info, e.g. which indices of the output correspond to the expected answer. The batch level info is probably harder to factor out.

File: composer/metrics/nlp.py [hunk 2]
```diff
@@ -196,6 +196,10 @@ def compute(self) -> Tensor:
 
 class InContextLearningMetric(Metric):
 
+    def __init__(self, *args, **kwargs):
+        super.__init__(*args, **kwargs)
```
Comment (bmosaicml) [inline composer/metrics/nlp.py:200]: Should it be `super().__init__` ?

File: composer/metrics/nlp.py [hunk 3]
```diff
@@ -196,6 +196,10 @@ def compute(self) -> Tensor:
 
 class InContextLearningMetric(Metric):
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (bmosaicml) [inline composer/metrics/nlp.py:200]: Added missing parentheses

File: composer/models/huggingface.py [hunk 4]
```diff
@@ -532,12 +531,8 @@ def get_metrics(self, is_train: bool = False) -> Dict[str, Metric]:
         return metrics if metrics else {}
 
     def update_metric(self, batch: Any, outputs: Any, metric: Metric) -> None:
-        if isinstance(metric, InContextLearningQAAccuracy):
-            assert self.labels is not None
+        if getattr(metric, 'needs_batch', False):
             metric.update(batch=batch, outputs=outputs, labels=self.labels)  # pyright: ignore [reportGeneralTypeIssues]
```
Comment (dakinggg) [inline composer/models/huggingface.py:535]: can you see if this type ignore can be removed now?

**Review by dakinggg — APPROVED** [review]

**Review by dakinggg — CHANGES_REQUESTED** [review]
Blocking because we found a bug.

File: composer/metrics/nlp.py [hunk 5]
```diff
@@ -328,9 +332,23 @@ def __init__(self, dist_sync_on_step: bool = False):
...
+               labels: Optional[torch.Tensor] = None,
+               outputs: Optional[torch.Tensor] = None):
+        if output_logits is not None:
+            warnings.warn(
+                ('`output_logits` has been deprecated and will be removed in a future release'),
+                DeprecationWarning,
+            )
+            outputs = output_logits
+
+        assert labels is not None
```
Comment (dakinggg) [inline composer/metrics/nlp.py:347]: this should raise with clear error if not none, rather than just assert

File: composer/metrics/nlp.py [hunk 6]
```diff
@@ -196,6 +196,10 @@ def compute(self) -> Tensor:
 
 class InContextLearningMetric(Metric):
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.needs_batch = True
+
     def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor):
```
Comment (dakinggg) [inline composer/metrics/nlp.py:203]: signature of the base class needs to change as well

File: composer/metrics/nlp.py [hunk 7]
```diff
@@ -328,9 +332,23 @@ def __init__(self, dist_sync_on_step: bool = False):
         self.add_state('correct', default=torch.tensor(0.), dist_reduce_fx='sum')
         self.add_state('total', default=torch.tensor(0.), dist_reduce_fx='sum')
 
-    def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor):
+    def update(self,
+               batch: dict,
+               output_logits: Optional[torch.Tensor] = None,
+               labels: Optional[torch.Tensor] = None,
+               outputs: Optional[torch.Tensor] = None):
+        if output_logits is not None:
```
Comment (dakinggg) [inline composer/metrics/nlp.py:340]: please pull this arg checking logic out into a helper function that returns the correct arguments so that the logic only exists in one place

**Review by bmosaicml — APPROVED** [review]

File: composer/metrics/nlp.py [hunk 8]
```diff
@@ -196,7 +196,15 @@ def compute(self) -> Tensor:
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.needs_batch = True
+
+    def update(self,
+               batch: dict,
+               output_logits: Optional[torch.Tensor] = None,
+               labels: Optional[torch.Tensor] = None,
+               outputs: Optional[torch.Tensor] = None):
         """Abstract interface for computing an in-context learning metrics.
```
Comment (dakinggg) [inline composer/metrics/nlp.py:208]: Leave a comment in the docstring that the args are changing and `output_logits` will be removed in v0.21

File: composer/metrics/nlp.py [hunk 9]
```diff
@@ -210,6 +218,27 @@ def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor)
         """
         raise NotImplementedError
 
+    @classmethod
+    def rename_args(
+            cls,
```
Comment (dakinggg) [inline composer/metrics/nlp.py:223]: since you don't use `cls`, this should be `@staticmethod`, yeah?

File: composer/metrics/nlp.py [hunk 10]
```diff
@@ -210,6 +218,27 @@ def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor)
...
+            batch: dict,
+            output_logits: Optional[torch.Tensor] = None,
+            labels: Optional[torch.Tensor] = None,
+            outputs: Optional[torch.Tensor] = None) -> Tuple[dict, Optional[torch.Tensor], torch.Tensor, torch.Tensor]:
+        if output_logits is not None:
+            warnings.warn(
+                ('`output_logits` has been renamed to `outputs   and will be removed in a future release'),
```
Comment (dakinggg) [inline composer/metrics/nlp.py:230]:
```suggestion
                ('`output_logits` has been renamed to `outputs   and will be removed in v0.21'),
```

File: composer/metrics/nlp.py [hunk 11]
```diff
@@ -210,6 +218,27 @@ def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor)
...
+                DeprecationWarning,
+            )
+            outputs = output_logits
+
+        if labels is None:
+            raise ValueError('`labels` cannot be None')
+        if outputs is None:
+            raise ValueError('`outputs` cannot be None')
+
+        return batch, output_logits, labels, outputs
```
Comment (dakinggg) [inline composer/metrics/nlp.py:240]: this should just return the three args we actually need, not `output_logits`

File: composer/metrics/nlp.py [hunk 12]
```diff
@@ -266,9 +299,8 @@ def replace_underscore(text: str) -> str:
 
         return white_space_fix(remove_articles(handle_punc(lower(replace_underscore(answer))))).strip()
 
-    def update(self, outputs: List[str], labels: List[List[str]], batch: Optional[Dict[str, Any]] = None):
-        if batch is None:
-            batch = {}
+    # TODO: rearrange these? why could batch be optional?
```
Comment (maxisawesome) [inline composer/metrics/nlp.py:302]: @dakinggg what do you think about making this change now? everywhere else has batch first

Comment (dakinggg) [reply composer/metrics/nlp.py:301]: wow thats annoying. hm...technically thats breaking, but maybe making the signature correct with the base class is more important and its unlikely to actually break anyone... @mvpatel2000 wdyt?

Comment (dakinggg) [reply composer/metrics/nlp.py:301]: @maxisawesome I vote we leave as is for now, and when we actually remove `output_logits`, we make the breaking change here too so that it matches the base class signature. and then call it out in the release notes.

File: composer/metrics/nlp.py [hunk 13]
```diff
@@ -210,6 +221,27 @@ def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor)
...
+    @staticmethod
+    def rename_args(batch: dict,
+                    output_logits: Optional[torch.Tensor] = None,
+                    labels: Optional[torch.Tensor] = None,
+                    outputs: Optional[torch.Tensor] = None) -> Tuple[dict, torch.Tensor, torch.Tensor]:
+        if outputs is not None and output_logits is not None:
+            raise ValueError('Cannot use both `outputs` and `output_logits`')
+        if output_logits is not None:
+            warnings.warn(
+                ('`output_logits` has been renamed to `outputs` and will be removed in a future release'),
```
Comment (dakinggg) [inline composer/metrics/nlp.py:233]:
```suggestion
                ('`output_logits` has been renamed to `outputs` and will be removed in v0.21'),
```

File: composer/metrics/nlp.py [hunk 14]
```diff
@@ -424,6 +473,7 @@ def __init__(self, dist_sync_on_step: bool = False, n_buckets: int = 10):
         self.add_state('bucket_correct', default=torch.zeros(n_buckets), dist_reduce_fx='sum')
 
     def update(self, batch: dict, output_logits: torch.Tensor, labels: torch.Tensor):
+        # TODO: fix
```
Comment (dakinggg) [inline composer/metrics/nlp.py:476]: what is going on here?

File: composer/metrics/nlp.py (hunk 12 above)
Comment (maxisawesome) [reply composer/metrics/nlp.py:301]: Okay, that works

File: composer/metrics/nlp.py (hunk 14 above)
Comment (maxisawesome) [reply composer/metrics/nlp.py:476]: removed

File: composer/metrics/nlp.py [hunk 15]
```diff
@@ -266,9 +298,7 @@ def replace_underscore(text: str) -> str:
 
         return white_space_fix(remove_articles(handle_punc(lower(replace_underscore(answer))))).strip()
 
-    def update(self, outputs: List[str], labels: List[List[str]], batch: Optional[Dict[str, Any]] = None):
-        if batch is None:
-            batch = {}
+    def update(self, outputs: List[str], labels: List[List[str]], batch: Dict[str, Any]):
```
Comment (dakinggg) [inline composer/metrics/nlp.py:301]: Is the batch not optional? it seems optional cause everything accessed on it is defaulted, and you don't have to specify any of those things

**Review by dakinggg — APPROVED** [review]
One comment, but otherwise LGTM

File: composer/metrics/nlp.py (hunk 15 above)
Comment (maxisawesome) [reply composer/metrics/nlp.py:301]: Only place it's called is in our original update_metric function which will always pass it a batch. Based on the code though, it's not optional (can't call .get on None)
I will note that this is a prime example jamming stuff in the batch that really is used in metric computation.

Comment (dakinggg) [reply composer/metrics/nlp.py:301]: well it was optional cause it used to do
```
if batch is None:
    batch = {}
```

Comment (maxisawesome) [reply composer/metrics/nlp.py:301]: we decided to fix it in a later release
